util/lsp/src/gotodefinition.rs

```diff
@@ -0,0 +1,167 @@
+//! Implementation of the goto-definition functionality of the LSP server
+
+use tower_lsp::{jsonrpc, lsp_types::*};
+
+use super::conversion::*;
+use super::server::*;
+use query::*;
+
+pub async fn goto_definition(
+    server: &Server,
+    params: GotoDefinitionParams,
+) -> jsonrpc::Result<Option<GotoDefinitionResponse>> {
+    let pos_params = params.text_document_position_params;
+    let text_document = pos_params.text_document;
+
+    server
+        .client
+        .log_message(MessageType::INFO, format!("GotoDefinition request: {}", text_document.uri))
+        .await;
+
+    let pos = pos_params.position;
+    let db = server.database.read().await;
+    let index = db.get(&text_document.uri).unwrap();
+    let info =
+        index.location_to_index(pos.from_lsp()).and_then(|idx| index.hoverinfo_at_index(idx));
+    let res = info.and_then(|info| info_to_jump(&index, info));
```

Comment:
I think decomposing this as `to_jump_content(info.definition_site(), index)` would be clearer, where `definition_site` is a method that just extracts the `Option<(Url, Span)>`.

Reply:
I have rewritten the implementation and removed some redundance. Take a look and tell me if you like the new decomposition better :)